File: Intatis-Windows/src/Intatis.Shared/ConversationSession.cs
using System.Diagnostics;

namespace Intatis.Windows.Shared;

public sealed class ConversationSession
{
    private readonly OpenAIClient _client;
    private readonly List<IntatisMessage> _messages = new();

    public IReadOnlyList<IntatisMessage> Messages => _messages.AsReadOnly();

    public ConversationSession(IntatisConfig config)
    {
        _client = new OpenAIClient(config);
        _messages.Add(new IntatisMessage(MessageRole.System,
            "You are Intatis (Windows). Provide short, practical responses."));
    }

    public async Task<(IntatisMessage Message, TimeSpan Latency, string? Usage)> SendUserMessageAsync(
        string userText,
        string? model = null,
        string? reasoning = null,
        IReadOnlyList<ChatAttachment>? attachments = null,
        CancellationToken cancellationToken = default)
    {
        var user = new IntatisMessage(MessageRole.User, userText);
        _messages.Add(user);

        var (text, latency, usage) = await _client.SendAsync(
            _messages,
            model,
            reasoning,
            attachments,
            cancellationToken);
        var assistant = new IntatisMessage(MessageRole.Assistant, text);
        _messages.Add(assistant);
        return (assistant, latency, usage);
    }

    public void Clear()
    {
        var system = _messages.FirstOrDefault(m => m.Role == MessageRole.System);
        _messages.Clear();
        if (system is not null)
            _messages.Add(system);
    }
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/session/ConversationSession.kt
package com.intatis.shared.session

import com.intatis.shared.attachments.ImageAttachment
import com.intatis.shared.model.IntatisConfig
import com.intatis.shared.model.IntatisMessage
import com.intatis.shared.model.MessageRole
import com.intatis.shared.provider.OpenAIClient
import java.time.Duration

class ConversationSession(config: IntatisConfig) {
    private val client = OpenAIClient(config)
    private val messages = mutableListOf<IntatisMessage>()

    init {
        messages.add(
            IntatisMessage(
                role = MessageRole.SYSTEM,
                content = "You are Intatis (Android). Provide short, practical responses.",
            ),
        )
    }

    fun clear() {
        val system = messages.firstOrNull { it.role == MessageRole.SYSTEM }
        messages.clear()
        if (system != null) messages.add(system)
    }

    suspend fun sendUserMessageAsync(
        userText: String,
        model: String?,
        reasoning: String?,
        attachments: List<ImageAttachment>,
    ): Triple<IntatisMessage, Duration, String?> {
        messages.add(IntatisMessage(role = MessageRole.USER, content = userText))
        val result = client.sendAsync(messages, model, reasoning, attachments)
        val assistant = IntatisMessage(role = MessageRole.ASSISTANT, content = result.text)
        messages.add(assistant)
        return Triple(assistant, Duration.ofMillis(result.latencyMs), result.usage)
    }
}
